packages/datetime/src/dateRangeInput.tsx

```diff
@@ -66,6 +73,28 @@ export interface IDateRangeInputProps extends IDatePickerBaseProps, IProps {
     onChange?: (selectedRange: DateRange) => void;
 
     /**
+     * Called when the user finishes typing in a new date and the date causes an error state.
+     * If the date is invalid, `new Date(undefined)` will be returned for the corresponding
+     * boundary of the date range.
+     * If the date is out of range, the out-of-range date will be returned for the corresponding
+     * boundary of the date range (`onChange` is not called in this case).
+     */
+    onError?: (errorRange: DateRange) => void;
+
+    /**
+     * The error message to display when the date selected is out of range.
+     * @default "Out of range"
+     */
+    outOfRangeMessage?: string;
+
+    /**
+     * The error message to display when the selected dates overlap.
+     * This can only happen when typing dates in the input field.
+     * @default "Overlapping dates"
+     */
+    overlappingDatesMessage?: string;
```

Comment:
I've thought a lot about this, and I feel strongly that an overlapping-dates error is the right way to go. Some designers and engineers I spoke to over the past few weeks each offered that when you type in a date for a particular boundary, you presumably want that date in particular to stick, to have priority, to matter more than what was already there—not to swap on you if it happens to overlap the other date. 

A useful example: imagine there's a date range `[A, B]` already specified in the two fields. I want to specify a new range `[C, D]` later than the current range, where both new bounds will be chronologically after the current end date `B`. I choose to type the dates into the fields, rather than specifying them via clicking.

**Without an overlapping-dates error message:**
1. I type my new start date `C`, which falls after my current end date `B`.
1. I <kbd>Tab</kbd> over to the end field, and the start date I just typed _follows me_ there (because the component auto-swapped the dates on blur to prevent overlap). I now have a range `[B, C]` in my two fields.
1. Probably confused, I now do one of two things:
    1. _Option 1 (6 total interactions):_
        1. I <kbd>Shift + Tab</kbd> back to the start field.
        1. I re-type the same start date `C`. Now my range is `[C, C]`. If `allowSingleDayRange=false`, my end date will auto-clear.
        1. I `Tab` over to the end field _again_.
        1. I type my new end date `D`.
    1. _Option 2 (5 total interactions):_
        1. I type my new end date `D` into the end field.
        1. I <kbd>Shift + Tab</kbd> back to the start field.
        1. I input the new start date `C`.

**With an overlapping-dates error message:**
1. I type my new start date `C`. I see an overlapping-dates error message in the end field.
2. I <kbd>Tab</kbd> over to the end field.
3. I type my new end date `D`.

The former approach involves confusing magic and balloons the number of interactions required to type a range. The latter keeps things quick and clear.